File: app/src/main/java/com/qinshou/rcvbaseadapterdemo/holder/BaseViewHolder.java
package com.qinshou.rcvbaseadapterdemo.holder;

import android.content.Context;
import android.graphics.Bitmap;
import android.support.annotation.ColorInt;
import android.support.annotation.DrawableRes;
import android.support.annotation.StringRes;
import android.support.v7.widget.RecyclerView;
import android.util.SparseArray;
import android.view.View;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.TextView;

/**
 * Description:RecyclerView 的 ViewHolder 的基类,可以封装各种通用的方法
 * 比如 View 的背景设置,是否选中的状态设置
 * TextView 的文字设置,字体大小设置,字体颜色设置
 * ImageView 的图片设置,当然,如果一般是网络加载的话,需要获得 ImageView 自己异步加载,也可以在该类中内置 Glide 等第三方库异步加载网络图片
 * 设置 OnClickListener,OnLongClickListener 等
 * Created by 禽兽先生
 * Created on 2017/11/22
 */

public class BaseViewHolder extends RecyclerView.ViewHolder {
    private Context context;
    private View itemView;
    private SparseArray<View> viewSparseArray;

    public BaseViewHolder(Context context, View itemView) {
        super(itemView);
        this.context = context;
        this.itemView = itemView;
        this.viewSparseArray = new SparseArray<View>();
    }

    public View getItemView() {
        return itemView;
    }

    public <T extends View> T findViewById(int viewId) {
        View view = viewSparseArray.get(viewId);
        if (view == null) {
            view = itemView.findViewById(viewId);
            viewSparseArray.put(viewId, view);
        }
        return (T) view;
    }

    /**
     * Description:设置控件是否可见
     * Date:2018/3/9
     */
    public BaseViewHolder setVisibility(int viewId, int visibility) {
        View mView = findViewById(viewId);
        if (mView != null) {
            mView.setVisibility(visibility);
        }
        return this;
    }

    /**
     * Description:设置 TextView 文本
     * Date:2018/3/9
     */
    public BaseViewHolder setTvText(int viewId, String text) {
        TextView mTextView = findViewById(viewId);
        if (mTextView != null) {
            mTextView.setText(text);
        }
        return this;
    }

    /**
     * Description:设置 TextView 文本
     * Date:2018/3/9
     */
    public BaseViewHolder setTvText(int viewId, @StringRes int resId) {
        TextView mTextView = findViewById(viewId);
        if (mTextView != null) {
            mTextView.setText(context.getResources().getString(resId));
        }
        return this;
    }

    /**
     * Description:设置 TextView 文本
     * Date:2018/3/9
     */
    public BaseViewHolder setTvText(int viewId, CharSequence text) {
        TextView mTextView = findViewById(viewId);
        if (mTextView != null) {
            mTextView.setText(text);
        }
        return this;
    }

    /**
     * Description:设置 Button 文本
     * Date:2018/3/9
     */
    public BaseViewHolder setBtnText(int viewId, String text) {
        Button mButton = findViewById(viewId);
        if (mButton != null) {
            mButton.setText(text);
        }
        return this;
    }

    /**
     * Description:设置 Button 文本
     * Date:2018/3/9
     */
    public BaseViewHolder setBtnText(int viewId, @StringRes int resId) {
        Button mButton = findViewById(viewId);
        if (mButton != null) {
            mButton.setText(context.getResources().getString(resId));
        }
        return this;
    }

    /**
     * Description:设置 ImageView 图片
     * Date:2018/3/9
     */
    public BaseViewHolder setIvImage(int viewId, Bitmap bitmap) {
        ImageView mImageView = findViewById(viewId);
        if (mImageView != null) {
            mImageView.setImageBitmap(bitmap);
        }
        return this;
    }

    /**
     * Description:设置 ImageView 图片
     * Date:2018/3/9
     */
    public BaseViewHolder setIvImage(int viewId, @DrawableRes int resId) {
        ImageView mImageView = findViewById(viewId);
        if (mImageView != null) {
            mImageView.setImageResource(resId);
        }
        return this;
    }

    /**
     * Description:设置 View 的点击事件监听器
     * Date:2018/3/9
     */
    public BaseViewHolder setOnClickListener(int viewId, View.OnClickListener onClickListener) {
        View mView = findViewById(viewId);
        if (mView != null) {
            mView.setOnClickListener(onClickListener);
        }
        return this;
    }

    /**
     * Description:设置 View 的 selected 状态
     * Date:2018/3/9
     */
    public BaseViewHolder setSelected(int viewId, boolean selected) {
        View mView = findViewById(viewId);
        if (mView != null) {
            mView.setSelected(selected);
        }
        return this;
    }

    /**
     * Description:设置 View 的背景色
     * Date:2018/3/9
     */
    public BaseViewHolder setBackgroundColor(int viewId, @ColorInt int color) {
        View mView = findViewById(viewId);
        if (mView != null) {
            mView.setBackgroundColor(color);
        }
        return this;
    }

    /**
     * Description:设置 View 的背景
     * Date:2018/3/9
     */
    public BaseViewHolder setBackgroundResource(int viewId, @DrawableRes int resId) {
        View mView = findViewById(viewId);
        if (mView != null) {
            mView.setBackgroundResource(resId);
        }
        return this;
    }
}

File: rcvbaseadapter/src/main/java/com/qinshou/rcv/viewholder/BaseViewHolder.kt
package com.qinshou.rcv.viewholder

import android.content.Context
import android.graphics.Bitmap
import android.graphics.drawable.Drawable
import android.util.SparseArray
import android.view.View
import android.widget.Button
import android.widget.ImageView
import android.widget.TextView
import androidx.annotation.ColorInt
import androidx.annotation.DrawableRes
import androidx.annotation.StringRes
import androidx.recyclerview.widget.RecyclerView

/**
 * Author: MrQinshou
 * Email:cqflqinhao@126.com
 * Date: 2021/4/14 10:37
 * Description:The basic of view holder,developers no longer need to care about the implementation
 * of the ViewHolder
 */
class BaseViewHolder(val context: Context, view: View) : RecyclerView.ViewHolder(view) {
    private val viewSparseArray: SparseArray<View> = SparseArray()

    /**
     * Author: MrQinshou
     * Email:cqflqinhao@126.com
     * Date:2021/4/14 18:13
     * Description:Return a view by id.
     */
    fun <T : View?> findViewById(viewId: Int): T? {
        var view: View? = viewSparseArray.get(viewId)
        if (view == null) {
            view = itemView.findViewById(viewId)
            viewSparseArray.put(viewId, view)
        }
        return view as T
    }

    /**
     * Author: MrQinshou
     * Email:cqflqinhao@126.com
     * Date:2021/4/14 18:13
     * Description:Return a text view by id.
     */
    fun getTextView(textViewId: Int): TextView? {
        var textView: TextView? = viewSparseArray.get(textViewId) as TextView
        if (textView == null) {
            textView = itemView.findViewById(textViewId)
            viewSparseArray.put(textViewId, textView)
        }
        return textView
    }

    /**
     * Author: MrQinshou
     * Email:cqflqinhao@126.com
     * Date:2021/4/14 18:13
     * Description:Return a image view by id.
     */
    fun getImageView(imageViewId: Int): ImageView? {
        var imageView: ImageView? = viewSparseArray.get(imageViewId) as ImageView
        if (imageView == null) {
            imageView = itemView.findViewById(imageViewId)
            viewSparseArray.put(imageViewId, imageView)
        }
        return imageView
    }

    /**
     * Author: MrQinshou
     * Email:cqflqinhao@126.com
     * Date:2021/4/14 18:13
     * Description:Set the visibility of view.
     */
    fun setVisibility(viewId: Int, visibility: Int): BaseViewHolder {
        val view = findViewById<View>(viewId)
        if (view != null) {
            view.visibility = visibility
        }
        return this
    }

    /**
     * Author: MrQinshou
     * Email:cqflqinhao@126.com
     * Date:2021/4/14 18:13
     * Description:Set the text of text view.
     */
    fun setTvText(viewId: Int, text: CharSequence?): BaseViewHolder {
        val textView = findViewById<TextView>(viewId)
        if (textView != null) {
            textView.text = text
        }
        return this
    }

    /**
     * Author: MrQinshou
     * Email:cqflqinhao@126.com
     * Date:2021/4/14 18:13
     * Description:Set the text of text view.
     */
    fun setTvText(viewId: Int, @StringRes resId: Int): BaseViewHolder {
        findViewById<TextView>(viewId)?.text = context.resources.getString(resId)
        return this
    }

    /**
     * Author: MrQinshou
     * Email:cqflqinhao@126.com
     * Date:2021/4/14 18:13
     * Description:Set the text of button.
     */
    fun setBtnText(viewId: Int, text: CharSequence?): BaseViewHolder {
        val button = findViewById<Button>(viewId)
        if (button != null) {
            button.text = text
        }
        return this
    }

    /**
     * Author: MrQinshou
     * Email:cqflqinhao@126.com
     * Date:2021/4/14 18:13
     * Description:Set the text of button.
     */
    fun setBtnText(viewId: Int, @StringRes resId: Int): BaseViewHolder {
        findViewById<Button>(viewId)?.text = context.resources.getString(resId)
        return this
    }

    /**
     * Author: MrQinshou
     * Email:cqflqinhao@126.com
     * Date:2021/4/14 18:13
     * Description:Set the img of image view.
     */
    fun setIvImage(viewId: Int, @DrawableRes resId: Int): BaseViewHolder {
        findViewById<ImageView>(viewId)?.setImageResource(resId)
        return this
    }

    /**
     * Author: MrQinshou
     * Email:cqflqinhao@126.com
     * Date:2021/4/14 18:13
     * Description:Set the img of image view.
     */
    fun setIvImage(viewId: Int, bitmap: Bitmap?): BaseViewHolder {
        findViewById<ImageView>(viewId)?.setImageBitmap(bitmap)
        return this
    }

    /**
     * Author: MrQinshou
     * Email:cqflqinhao@126.com
     * Date:2021/4/14 18:13
     * Description:Set the img of image view.
     */
    fun setIvImage(viewId: Int, drawable: Drawable?): BaseViewHolder {
        findViewById<ImageView>(viewId)?.setImageDrawable(drawable)
        return this
    }

    /**
     * Author: MrQinshou
     * Email:cqflqinhao@126.com
     * Date:2021/4/14 18:13
     * Description:Set the listener of click of view.
     */
    fun setOnClickListener(viewId: Int, onClickListener: View.OnClickListener?): BaseViewHolder {
        findViewById<View>(viewId)?.setOnClickListener(onClickListener)
        return this
    }

    /**
     * Author: MrQinshou
     * Email:cqflqinhao@126.com
     * Date:2021/4/14 18:13
     * Description:Set the state of selected of view.
     */
    fun setSelected(viewId: Int, selected: Boolean): BaseViewHolder {
        val view = findViewById<View>(viewId)
        if (view != null) {
            view.isSelected = selected
        }
        return this
    }

    /**
     * Author: MrQinshou
     * Email:cqflqinhao@126.com
     * Date:2021/4/14 18:13
     * Description:Set the background of view.
     */
    fun setBackgroundColor(viewId: Int, @ColorInt color: Int): BaseViewHolder {
        findViewById<View>(viewId)?.setBackgroundColor(color)
        return this
    }

    /**
     * Author: MrQinshou
     * Email:cqflqinhao@126.com
     * Date:2021/4/14 18:13
     * Description:Set the background of view.
     */
    fun setBackgroundResource(viewId: Int, @DrawableRes resId: Int): BaseViewHolder {
        findViewById<View>(viewId)?.setBackgroundResource(resId)
        return this
    }
}